signer/src/lib.rs

```diff
@@ -334,18 +335,112 @@ pub fn verify_signature(
+///
+/// * `signature` - RSV format signature or BLS signature
+/// * `cbor_buffer` - the CBOR transaction to verify the signature against
+///
+pub fn verify_signature(
+    signature: &Signature,
+    cbor_buffer: &CborBuffer,
+) -> Result<bool, SignerError> {
+    let result = match signature {
+        Signature::SignatureSECP256K1(sig_secp256k1) => {
+            verify_secp256k1_signature(sig_secp256k1, cbor_buffer)?
+        }
+        Signature::SignatureBLS(sig_bls) => verify_bls_signature(sig_bls, cbor_buffer)?,
+    };
+
+    Ok(result)
+}
+
+fn extract_from_pub_key_from_message(
+    cbor_message: &CborBuffer,
+) -> Result<bls_signatures::PublicKey, SignerError> {
+    let message = transaction_parse(cbor_message, true)?;
+
+    let unsigned_message_api = message.get_message();
+    let from_address = Address::from_str(&unsigned_message_api.from.to_string())?;
+
+    let pk = bls_signatures::PublicKey::from_bytes(&from_address.payload())?;
+
+    Ok(pk)
+}
+
+pub fn verify_aggregated_signature(
+    signature: &SignatureBLS,
+    cbor_messages: &[CborBuffer],
+) -> Result<bool, SignerError> {
+    let sig = bls_signatures::Signature::from_bytes(signature.as_ref())?;
+
+    // Get public keys from message
+    let tmp: Result<Vec<_>, SignerError> = cbor_messages
+        .into_iter()
+        .map(|cbor_message| extract_from_pub_key_from_message(cbor_message))
+        .collect();
+
+    let pks = match tmp {
+        Ok(public_keys) => public_keys.to_owned(),
+        Err(_) => {
+            return Err(SignerError::GenericString(
+                "Invalid public key extracted from message".to_string(),
+            ));
+        }
+    };
+
+    // Hashes
+    let hashes: Vec<_> = cbor_messages
+        .par_iter()
+        .map(|cbor_message| bls_signatures::hash(cbor_message.as_ref()))
+        .collect::<Vec<_>>();
+
+    return Ok(bls_signatures::verify(&sig, &hashes, pks.as_slice()));
+}
+
 #[cfg(test)]
 mod tests {
     use crate::api::{MessageTxAPI, UnsignedMessageAPI};
+    use crate::signature::{Signature, SignatureBLS};
     use crate::utils::{from_hex_string, to_hex_string};
     use crate::{
         key_derive, key_derive_from_seed, key_generate_mnemonic, key_recover, transaction_parse,
-        transaction_sign_raw, verify_signature, CborBuffer, Mnemonic, PrivateKey,
+        transaction_serialize, transaction_sign_bls_raw, transaction_sign_raw,
+        verify_aggregated_signature, verify_signature, CborBuffer, Mnemonic, PrivateKey,
     };
     use bip39::{Language, Seed};
     use forest_encoding::to_vec;
     use std::convert::TryFrom;
 
+    use crate::utils;
+    use bls_signatures;
+    use bls_signatures::Serialize;
+    use forest_address::Address;
+    use rand::{Rng, SeedableRng};
+    use rand_xorshift::XorShiftRng;
+    use rayon::prelude::*;
+
+    const BLS_PUBKEY: &str = "ade28c91045e89a0dcdb49d5ed0d62a4f02d78a96dbd406a4f9d37a1cd2fb5c29058def79b01b4d1556ade74ffc07904";
+    // FIXME! Might be invalid
```

Comment:
is this something serious to correct or verify?

Reply:
This one was generated using `bls-signatures` so I think we are good. But it wasn't working with the bls javascript lib I tried in the wasm tests.

I don't think it is relevant anymore. We discussed that the javascript lib I was using was most certainly using an other standard and we should instead generate test cases from dig's lib.